Add AWS STS assume role authentication capability

hackgnar opened the pull request:
This adds the capability to authenticate to AWS via STS assume role.  This functionality grabs temporary security credentials via assume role and uses those for connections.  As these credentials are temporary, there are also checks to repull new temporary credentials when old ones expire.

This allows for more secure connections via the OSQuery AWS module.  More information on AWS STS can be found here:
http://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

theopolis commented:
Awesome, thanks Ryan! FYI, I'm about to tag a 1.7.5 release. Unfortunately we wont be able to include this STS capability in that release since it requires modification to the build process.

theopolis commented:
Tagging this "In Progress" since I'm still blocked by some build configuration needs.

hackgnar commented:
@theopolis just checked this against master to make sure there were no merge conflicts with recent aws changes.  Its still gtg.

theopolis commented:
Right, it's blocked by the build hosts not having anything other than `core`, `kinesis`, and `firehose` installed. :)

theopolis commented:
Just to ping again, I should have an easier solution for managing dependencies soon that will unblock this.

hackgnar commented:
@theopolis @zwass while you guys are changing the build boxes for this PR, is there anything you guys want me to change in the code so its gtg when you guys are ready to merge it?  Changes such as:

1. Style changes
2. Optimizations
3. Proper coding procedures
4. etc

zwass commented on docs/wiki/deployment/aws-logging.md:
```diff
@@ -11,6 +11,10 @@ Some configuration is shared between the two plugins:
 --aws_region VALUE                      AWS region override
 --aws_secret_access_key VALUE           AWS secret access key override
+--aws_sts_arn_role VALUE                AWS STS assume role arn
+--aws_sts_region VALUE                  AWS STS assime role region
```
typo

zwass commented on docs/wiki/deployment/aws-logging.md:
```diff
@@ -21,6 +25,8 @@ When working with AWS, osquery will look for credentials and region configuratio
 4. `default` profile in the AWS config files
 5. Profile from the EC2 Instance Metadata Service
 
+All of the STS configuration flags are optional.  However, if `aws_sts_arn_role` is set, you can utilise temparary credentials via assume role with [AWS Security Token Service](http://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html).  
```
temparary -> temporary

zwass commented on osquery/logger/plugins/aws_util.h:
```diff
@@ -77,20 +79,42 @@ class OsqueryFlagsAWSCredentialsProvider
+  Aws::String sts_access_key_id = "";
+  Aws::String sts_secret_access_key = "";
+  Aws::String sts_session_token = "";
+  long long token_expire_time = 0;
```
`time_t` or `std::chrono` would be preferred.

zwass replied:
Actually it looks like we use `size_t` throughout the code base when we want unsigned values to hold times.

zwass commented on osquery/logger/plugins/aws_util.h:
```diff
@@ -77,20 +79,42 @@ class OsqueryFlagsAWSCredentialsProvider
 /**
+ * @brief AWS credentials provider that uses STS assume role auth
+ *
+ * This provider deligates temp AWS STS credentials via assume role
```
deligates -> delegates

zwass commented on osquery/logger/plugins/aws_util.h:
```diff
@@ -77,20 +79,42 @@ class OsqueryFlagsAWSCredentialsProvider
+ * @brief AWS credentials provider that uses STS assume role auth
+ *
+ * This provider deligates temp AWS STS credentials via assume role
+ * for and AWS arn.
```
and -> an?

zwass commented on osquery/logger/plugins/aws_util.cpp:
```diff
@@ -179,10 +191,46 @@ OsqueryFlagsAWSCredentialsProvider::GetAWSCredentials() {
+    sts_r.SetRoleSessionName(FLAGS_aws_sts_session_name);
+    sts_r.SetDurationSeconds(FLAGS_aws_sts_timeout);
+    // Pull our STS creds
+    auto sts_outcome = sts_client_->AssumeRole(sts_r);
```
I know these are nasty templated types, but I would prefer to see them (rather than `auto`) so it's easier to look up the docs in the future.

zwass commented on osquery/logger/plugins/aws_util.cpp:
```diff
@@ -179,10 +191,46 @@ OsqueryFlagsAWSCredentialsProvider::GetAWSCredentials() {
+    // Pull our STS creds
+    auto sts_outcome = sts_client_->AssumeRole(sts_r);
+    if (sts_outcome.IsSuccess()) {
+      auto sts_result = sts_outcome.GetResult();
```
also here

hackgnar commented:
Awesome!  Thx @zwass !  Ill get those changes in asap.

hackgnar commented:
@zwass all of your requests, fixes, etc have been added.  Let me know if there is anything else to change.  Thanks again for all the help and suggestions!

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

zwass commented:
I am good with this. I think we are waiting on the build system update from @theopolis.

theopolis commented:
Ok, now we're ready to rock and roll with the build. This will be a good first test!

To rebase this for the build, update the formula line: https://github.com/facebook/osquery/blob/master/tools/provision/formula/aws-sdk-cpp.rb#L28 to include the STS "module?".

Then add a `revision 1` under the `url` header: https://github.com/facebook/osquery/blob/master/tools/provision/formula/aws-sdk-cpp.rb#L6

That will cause our build hosts to compile AWS-SDK from source. Each build host should have the required build tools, but it will be interesting to see what happens.

theopolis commented:
To test locally, make the changes to the formula then run:
```
make distclean
DEBUG=True OSQUERY_BUILD_DEPS=True make deps
make
```

This will take a LONG time, but will build everything from source. This will provide an absolute "yes" if STS will work. And this time is an investment, so you wont have to wait as long if you plan on making additional formula changes.

To make extra-sure, do this in the `ubuntu16.04` vagrant target after having installed `make` and `python`.

hackgnar commented:
Ill update the build formula, try out your build methods and test a bit today.... Should have it done and gtg by end of day.  Thx guys, cant wait to get this feature upstream.

hackgnar commented:
Still need to test on OSX, but on a stock ubuntu 16.04 it builds with:
`OSQUERY_BUILD_DEPS=True make deps`
but does not build with just:
`make deps`

As for sending data to aws, there seems to be some http issues with the aws sdk lib that @zwass will likely run into as well.  I keep getting certificate errors when attempting to make kinesis connections:
`E0801 11:42:50.075052  5924 aws_util.cpp:173] Exception making HTTP request to url (https://kinesis.us-east-1.amazonaws.com): certificate verify failed`

I get the above error when attempting sts auth and with just standard kinesis auth with no sts.

hackgnar commented:
@theopolis did the same on OSX.  The build process for sts seems to be working flawlessly (great job on the build deps changes).  However, we are getting cert issues with aws connections.  Same as above:
`E0801 14:27:27.279168 2001006592 aws_util.cpp:173] Exception making HTTP request to url (https://sts.us-east-1.amazonaws.com): certificate verify failed`

@zwass if I get some time today, I'll try to figure out what is going wrong with the http lib.

zwass commented:
@hackgnar The `certificate verify failed` is a problem that came up for others as well. You'll need to rebase to pull in my recent PR (https://github.com/facebook/osquery/pull/2299), and use `--tls_server_certs` as @theopolis describes in chat (https://osquery.slack.com/archives/osx/p1469754492000089).

hackgnar commented:
@zwass @theopolis everything works perfect.  I was just missing the new addition of the run flag `--tls_server_certs`

Should be gtg for builds.  Thx for the help guys!

zwass commented:
ok to test

osqueryer commented:
:-1: The commit c1d82381ef0c27bc4f91b229f917538300452d87 (Job results: [2975](https://jenkins.osquery.io/job/osqueryPullRequestBuild/2975/)) failed one or more tests.

hackgnar commented:
@theopolis Would the above fails be caused by `make deps` instead of `OSQUERY_BUILD_DEPS=True make deps` on *nix distros?  I noticed that this morn when testing.

theopolis commented:
Right, the build hosts need to be "primed" with all of the dependency tools brought in by the from-source build. Give me a moment to schedule that.

theopolis commented:
ok to test

theopolis commented:
Hmm, this doesn't seem correct:
```
==> cmake .. -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_INSTALL_PREFIX=/usr/local/osquery/Cellar/aws-sdk-cpp/0.13.8_1 -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DCMAKE_BUILD_WITH_INSTALL_RPATH=1 -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=1 -DSTATIC_LINKING=1 -DNO_HTTP_CLIENT=1 -DMINIMIZE_SIZE=ON -DBUILD_SHARED_LIBS=OFF -DBUILD_ONLY=firehose;kinesis;sts
==> make
==> make install
Warning: tried to install empty array to /usr/local/osquery/Cellar/aws-sdk-cpp/0.13.8_1/lib
/usr/local/osquery/Cellar/aws-sdk-cpp/0.13.8_1: 217 files, 7.3M, built in 1 minute 52 seconds
```

osqueryer commented:
:-1: The commit c1d82381ef0c27bc4f91b229f917538300452d87 (Job results: [2978](https://jenkins.osquery.io/job/osqueryPullRequestBuild/2978/)) failed one or more tests.

hackgnar commented:
@theopolis please let me know if there is anything I can do to assist you on this build issue.

theopolis commented:
Found it, this is another oversight with build host dependency updating. Thanks for hanging in there @hackgnar, I'll get this fixed up shortly.

1. I'm going to add `DEBUG=1` to the build host scripts.
2. The actual issue is the dependency upgrade is using our runtime `gcc` instead of our runtime `clang` so the AWS-SDK so the built symbols have a C++ ABI-tag inserted.

theopolis commented:
ok to test

theopolis commented:
Right, so let's see if it still fails after you rebase.

osqueryer commented:
:-1: The commit c1d82381ef0c27bc4f91b229f917538300452d87 (Job results: [2982](https://jenkins.osquery.io/job/osqueryPullRequestBuild/2982/)) failed one or more tests.

osqueryer commented:
:-1: The commit 77c589c2d4227d34f692f101a14dbac992c7a1d9 (Job results: [2983](https://jenkins.osquery.io/job/osqueryPullRequestBuild/2983/)) failed one or more tests.

theopolis commented:
Looks good!

hackgnar commented:
Yep... looks like the ubuntu 16 build finished ok in jenkins but just hung on completion which caused the "1 failing check"

theopolis commented on docs/wiki/deployment/aws-logging.md:
```diff
@@ -11,6 +11,10 @@ Some configuration is shared between the two plugins:
 --aws_region VALUE                      AWS region override
 --aws_secret_access_key VALUE           AWS secret access key override
+--aws_sts_arn_role VALUE                AWS STS assume role arn
+--aws_sts_region VALUE                  AWS STS assume role region
```
Is this region required? Seems like STS is automatic but a region is optional. Could we use the `--aws_region` flag that already exists?

hackgnar replied:
I am fairly certain that if the aws region and the region for the sts assume role differ, this flag will be needed.  I can test it again.

theopolis replied:
How often is that the case? Can we get by without adding a STS region?

hackgnar replied:
Unfortunately, its the case for me most of the time.  Ill test it tomorrow to see if it works without specifying an assume role region but I think I have tried that before.

hackgnar replied:
@theopolis ya I just tested this.  The sts region flag is needed.  When my sts assume role is in another region than the region I used for initial login.  As I mentioned above, this is the case for my setup most of the time.

theopolis commented on osquery/logger/plugins/aws_util.cpp:
```diff
@@ -179,10 +191,47 @@ OsqueryFlagsAWSCredentialsProvider::GetAWSCredentials() {
+  if (token_expire_time <= seconds.count()) {
+    // Create and setup a STS client to pull our temp creds
+    VLOG(1) << "Generate new AWS STS credentials.";
+    initAwsSdk();
```
It seems like this method will be called several times during the application life. I don't think `initAwsSdk` is meant to be called more than once. /cc @zwass

hackgnar replied:
Ya, with the sts addition, it will get called every time the sts creds expire.  Ive had this running on several machines for ~30 days and everything is still trucking.

theopolis replied:
My worry is mostly esthetic, since the initialization is a call_once we shouldn't put it in a loop or callback if we can avoid it. ;)

hackgnar replied:
Looking at it now, but I believe the `OsquerySTSAWSCredentialsProvider::GetAWSCredentials()` is a callback that gets triggered when auth fails since it was set int he `AWSCredentialsProviderChain`.

Playing around with some code to only call `initAwsSdk` on its first call.  I'll add the change if everything pans out.

zwass replied:
`initAwsSdk` is only meant to be called once, but as @theopolis mentions, it's guarded by the `std::once_flag`. There aren't any negative effects of calling it multiple times, but it won't actually do anything after the first call.

theopolis commented on osquery/logger/plugins/aws_util.cpp:
```diff
@@ -179,10 +191,47 @@ OsqueryFlagsAWSCredentialsProvider::GetAWSCredentials() {
+    if (sts_outcome.IsSuccess()) {
+      Aws::STS::Model::AssumeRoleResult sts_result = sts_outcome.GetResult();
+      // Cache our credentials for later use
+      sts_access_key_id = sts_result.GetCredentials().GetAccessKeyId();
```
We could set these variables as private, since they are passed to `::AWSCredentials`, right?

theopolis commented on osquery/logger/plugins/aws_util.cpp:
```diff
@@ -179,10 +191,47 @@ OsqueryFlagsAWSCredentialsProvider::GetAWSCredentials() {
+  // Grab system time in seconds since epoch for token experation checks
+  auto now = std::chrono::system_clock::now();
+  auto epoch = now.time_since_epoch();
+  auto seconds = std::chrono::duration_cast<std::chrono::seconds>(epoch);
```
We have a `getUnixTime()` that should do the trick.

theopolis commented on osquery/logger/plugins/aws_util.cpp:
```diff
@@ -179,10 +191,47 @@ OsqueryFlagsAWSCredentialsProvider::GetAWSCredentials() {
+    // Pull our STS creds
+    Aws::STS::Model::AssumeRoleOutcome sts_outcome =
+        sts_client_->AssumeRole(sts_r);
+    if (sts_outcome.IsSuccess()) {
```
So if this fails, the previous tuple of (ID, secret, token) will be sent. I'm assuming that will fail, but perhaps a `LOG(ERROR)` would be help? Though if you cannot log because of bad credentials, then additional status logging may be moot.

hackgnar commented:
@theopolis Ill add your recommended changes asap.  Once again, thx for the suggestions.

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

hackgnar commented:
@theopolis @zwass , I have added the following changes based on the above recommendations:

1. Added a check to only call initAwsSdk once
2. Added error logging when STS cred pull fails
3. Moved STS id, key & token vars to be private
4. Switched from `chrono` to `osquery::getUnixTime`

Let me know if there is anything else I should change/add.  Once again, thanks for your guidance and recommendations.

osqueryer commented:
:no_good: The commit d9ffedba24716c9abd9018ed1701475305f06bd2 (Job results: [36](https://jenkins.osquery.io/job/osqueryPullRequestAudit/36/)) failed the code audit.

osqueryer commented:
:-1: The commit d9ffedba24716c9abd9018ed1701475305f06bd2 (Job results: [3053](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3053/)) failed one or more tests (**OS X/Linux**).
